confluentinc/ksql
# feat: Replay command topic to local file to backup KSQL Metastore

### Description 
Implements KLIP-31: Metastore Backups - https://github.com/confluentinc/ksql/pull/5741

Add a metastore backup service that replays the command_topic to a local file. This local file will be useful to users who desire to restore their command_topic in case of accidental topic disasters.

Two new configs are added to enable backups (see `KsqlConfig`):
- `ksql.enable.metastore.backup` to enable the metastore backup
- `ksql.metastore.backup.location` to specify the location of the backup

Two new classes are created:
- `CommandTopicBackup`
- `BackupReplayFile`

The `CommandTopicBackup` is the service used to backup command topic records through the `writeRecord` method. It is called by the `CommandTopic` class on each record read by the Kafka consumer (see `getRestoreCommands` and `getNewCommands` from the `CommandTopic`). The `CommandTopicBackup` class creates a new backup file (if none exists) or opens a current backup file used in a previous ksqlDB restart to update it with only new commands. It may also create a new backup file in case the record to write does not match the previous ksqlDB replayed records. This is how it handles command_topic deletion accidents.

The `BackupReplayFile` is used to read/write records to a local file, and used by the `CommandTopicBackup`. Each record is serialized/deserialized using the JSON serde, which is the same used by the command_topic. The record written to the file contains only the CommandId (key) and Command (value) separated by a colon. This formatted data in the file allows users to easily restore their command_topic by executing the following Kafka command:
```
$ kafka-console-producer --broker-list localhost:9092 --topic $COMMAND_TOPIC --property "parse.key=true" --property "key.separator=:" < $BACKUP_FILE
```

Other classes modified are:
- `CommandTopic`. Calls the `CommandTopicBackup` on each record read by the command topic consumer.
- `CommandStore`. Creates the `CommandTopicBackup` if enabled, and passes it to the `CommandTopic` class.

### Testing done 
- Added unit tests on the new and modified classes
- Verified manually by execution the following scenarios:
   1. Create 1 stream, stop ksqlDB, delete/restore the command topic, start ksqlDB, run `show streams` [OK]
   2. Create 1 stream, stop ksqlDB, delete the backup file, start ksqlDB, check the backup file is created again [OK]
   3. Create 2 streams, stop ksqlDB, delete 2nd stream from backup, start ksqlDB, check only 2nd stream is added back [OK]
   4. Create 2 streams, stop ksqlDB, delete 1st stream from backup, start ksqlDB, check a new backup is created [OK]

### Reviewer checklist
